### DISCUSSION TITLE
IncomingInterceptor and OutgoingInterceptor

### DISCUSSION THREAD

**Discussion by mrickly** [smallrye/smallrye-reactive-messaging#1964] [Q&A]
The naming in the ChannelDecorators doc is confusing me: 
 
In https://smallrye.io/smallrye-reactive-messaging/4.0.0/concepts/decorators/#intercepting-outgoing-messages, it says
`Smallrye Reactive Messaging provides a SubscriberDecorator by default to allow intercepting **outgoing** messages for a specific channel`. 
On the other hand, at the top of the page https://smallrye.io/smallrye-reactive-messaging/4.0.0/concepts/decorators/, it says `**@Incoming** channels and channels bound to an outbound connector are both Subscribers`. 

Also, why is there a default implementation of a SubscriberDecorator, but, apparently, not of a PublisherDecorator using `IncomingInterceptor`?

**Comment by scrocquesel** [comment]
Here is my understanding.

Interceptors decorate an extremity of a channel formed by an `@Incoming` and `@Outgoing`, not connectors. A connector can be seen as only half of a channel (the other half is talking with the remote broker). 

For eg. outbound connectors actually subscribe like `@Incoming` (and thus messages from the channel flow into the connector). So if you want to intercept a message, you will implement a `SubscriberDecorator`.
On the other hand, inbound connectors publish messages like `@Outgoing` (messages flow out of the connector into the channel). Messages can be intercepted with a `PublisherDecorator` (`isConnector = true`). To be exhaustive, messages will then flow to a `@Incoming` endpoint and can also be intercepted before reaching the endpoint with a  `SubscriberDecorator`. This time,  `isConnector` will be `false` for the given channel name.

As for the default implementation. I guess this is purely fo historical reason. Before  channel decorators, incoming messages (so outgoing message connector speaking) could be intercepted with a MessageConverter and `OutgoingInterceptor` is a proper replacement.

**Reply by mrickly** [reply]
Hi @scrocquesel, thank you for taking the time to share your thoughts on this. I have written concrete examples meanwhile:
- `KafkaIncomingDecorator implements PublisherDecorator` -> decorates messages arriving from incoming channels.
- `KafkaOutgoingDecorator implements SubscriberDecorator` -> decorates messages emitted to outgoing channels.
The implementations do what their names suggest, which is the opposite of what the interface names suggest (at least for my intuition).
But as you mentioned, the idea is probably that the messages are published from the incoming connector to the `@Incoming` method parameters and that the outgoing connector subscribes to an emitter. This will have to be confirmed by the people working on smallrye-reactive-messaging. Same for the history leading to the asymmetrical construction (existence of `OutgoingInterceptor` and `OutgoingInterceptorDecorator`, but not of `IncomingInterceptor` and `IncomingInterceptorDecorator`). 

Cheers.

Matthieu
